run/websockets/redis.js

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const {promisify} = require('util');
+
+// [START cloudrun_websockets_redis]
+const REDISHOST = process.env.REDISHOST || 'localhost';
+const REDISPORT = process.env.REDISPORT || 6379;
+
+// Instantiate the Redis client
+const redisClient = require('redis').createClient(REDISPORT, REDISHOST);
+// [END cloudrun_websockets_redis]
+
+// Set up to use promises and async/await for Redis methods
+const redisGet = promisify(redisClient.get).bind(redisClient);
+const redisExists = promisify(redisClient.exists).bind(redisClient);
+
+// Insert new messages into the Redis cache
+async function addMessageToCache(roomName, msg) {
+  let room = await getRoomFromCache(roomName);
+  if (room) {
+    // Update old room
+    room.messages.push(msg);
+  } else {
+    // Create a new room
+    room = {
+      room: roomName,
+      messages: [msg],
+    };
+  }
+  redisClient.set(roomName, JSON.stringify(room));
+  addMessageToDb(room);
+}
+
+// Insert messages into the database
+async function addMessageToDb(data) {
+  const room = messageDb.find(messages => messages.room === data.room);
+  if (room) {
+    // Update room in database
+    Object.assign(room, data);
```

Comment:
suggestion: I'm not following the separate responsibilities of `addMessageToCache` and `addMessageToDb`. It might help if `messageDb` were initialized first, and had a bit more context on why it exists alongside redis.